File: ros2-basic/helloworld/src/node_helloworld_cpp/src/node_helloworld.cpp
#include "rclcpp/rclcpp.hpp"

class NodeHelloworld : public rclcpp::Node
{
public:
    NodeHelloworld(std::string name) : Node(name)
    {
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&NodeHelloworld::timer_callback, this));
    }

private:
    void timer_callback()
    {
        RCLCPP_INFO(this->get_logger(), "Hello World C++: %d", ++count_);
    }

private:
    int count_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NodeHelloworld>("node_helloworld_cpp");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2-basic/helloworld/src/node_helloworld_py/node_helloworld_py/node_helloworld.py
import rclpy
from rclpy.node import Node


class NodeHelloworld(Node):
    def __init__(self, name: str):
        super().__init__(name)
        self.count_ = 0
        # 创建 1 秒周期的定时器
        self.timer_ = self.create_timer(1.0, self.timer_callback)

    def timer_callback(self):
        self.count_ += 1
        self.get_logger().info(f"Hello World Python: {self.count_}")


def main(args=None):
    rclpy.init(args=args)
    node = NodeHelloworld("node_helloworld_cpp")
    rclpy.spin(node)
    rclpy.shutdown()

